File: class144_Backtracking/LeetCode78_Subsets.java
import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode 78. 子集
 * 
 * 题目描述：
 * 给你一个整数数组 nums ，数组中的元素互不相同。返回该数组所有可能的子集（幂集）。
 * 解集不能包含重复的子集。你可以按任意顺序返回解集。
 * 
 * 示例：
 * 输入：nums = [1,2,3]
 * 输出：[[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
 * 
 * 输入：nums = [0]
 * 输出：[[],[0]]
 * 
 * 提示：
 * 1 <= nums.length <= 10
 * -10 <= nums[i] <= 10
 * nums 中的所有元素互不相同
 * 
 * 链接：https://leetcode.cn/problems/subsets/
 * 
 * 算法思路：
 * 1. 使用回溯算法生成所有可能的子集
 * 2. 对于每个元素，有两种选择：选择或不选择
 * 3. 通过递归和回溯生成所有满足条件的子集
 * 4. 每一步递归都将当前路径加入结果集
 * 
 * 剪枝策略：
 * 1. 可行性剪枝：通过起始索引避免重复选择元素
 * 2. 最优性剪枝：当已选择的元素个数等于数组长度时终止
 * 3. 约束传播：一旦某个元素被选择，后续只能选择后面的元素
 * 
 * 时间复杂度：O(n * 2^n)，其中n是数组长度，共有2^n个子集，每个子集需要O(n)时间复制
 * 空间复杂度：O(n)，递归栈深度和存储路径的空间
 * 
 * 工程化考量：
 * 1. 边界处理：处理空数组和单元素数组的特殊情况
 * 2. 参数验证：验证输入参数的有效性
 * 3. 性能优化：通过剪枝减少不必要的计算
 * 4. 内存管理：合理使用数据结构减少内存占用
 * 5. 可读性：添加详细注释和变量命名
 * 6. 异常处理：处理可能的异常情况
 * 7. 模块化设计：将核心逻辑封装成独立方法
 * 8. 可维护性：添加详细注释和文档说明
 */
public class LeetCode78_Subsets {
    
    /**
     * 生成数组的所有子集
     * 
     * @param nums 输入数组
     * @return 所有可能的子集
     */
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        
        // 边界条件检查
        if (nums == null) {
            return result;
        }
        
        List<Integer> path = new ArrayList<>();
        backtrack(nums, 0, path, result);
        return result;
    }
    
    /**
     * 回溯函数生成子集
     * 
     * @param nums 输入数组
     * @param start 当前起始索引
     * @param path 当前路径
     * @param result 结果列表
     */
    private void backtrack(int[] nums, int start, List<Integer> path, List<List<Integer>> result) {
        // 每一步都添加到结果中（空集也在其中）
        result.add(new ArrayList<>(path));
        
        // 从start开始遍历，避免重复
        for (int i = start; i < nums.length; i++) {
            // 选择当前元素
            path.add(nums[i]);
            
            // 递归处理下一个元素
            backtrack(nums, i + 1, path, result);
            
            // 回溯：撤销选择
            path.remove(path.size() - 1);
        }
    }
    
    /**
     * 解法二：使用位运算枚举所有可能的子集
     * 每个元素有两种状态：选择(1)或不选择(0)，共2^n种组合
     * 
     * @param nums 输入数组
     * @return 所有可能的子集
     */
    public List<List<Integer>> subsets2(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        
        // 边界条件检查
        if (nums == null) {
            return result;
        }
        
        int n = nums.length;
        
        // 枚举所有可能的子集（0到2^n-1）
        for (int mask = 0; mask < (1 << n); mask++) {
            List<Integer> subset = new ArrayList<>();
            
            // 根据位掩码构造子集
            for (int i = 0; i < n; i++) {
                // 检查第i位是否为1
                if ((mask & (1 << i)) != 0) {
                    subset.add(nums[i]);
                }
            }
            
            result.add(subset);
        }
        
        return result;
    }
    
    /**
     * 解法三：使用迭代法生成所有子集
     * 逐个添加元素，每次添加新元素时，将该元素添加到已有的所有子集中
     * 
     * @param nums 输入数组
     * @return 所有可能的子集
     */
    public List<List<Integer>> subsets3(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        
        // 边界条件检查
        if (nums == null) {
            return result;
        }
        
        // 初始化空集
        result.add(new ArrayList<>());
        
        // 逐个添加元素
        for (int num : nums) {
            int size = result.size();
            
            // 将当前元素添加到已有的所有子集中
            for (int i = 0; i < size; i++) {
                List<Integer> newSubset = new ArrayList<>(result.get(i));
                newSubset.add(num);
                result.add(newSubset);
            }
        }
        
        return result;
    }
    
    // 测试方法
    public static void main(String[] args) {
        LeetCode78_Subsets solution = new LeetCode78_Subsets();
        
        // 测试用例1
        int[] nums1 = {1, 2, 3};
        List<List<Integer>> result1 = solution.subsets(nums1);
        System.out.println("输入: nums = [1,2,3]");
        System.out.println("输出: " + result1);
        
        // 测试用例2
        int[] nums2 = {0};
        List<List<Integer>> result2 = solution.subsets(nums2);
        System.out.println("\n输入: nums = [0]");
        System.out.println("输出: " + result2);
        
        // 测试解法二
        System.out.println("\n=== 解法二测试 ===");
        List<List<Integer>> result3 = solution.subsets2(nums1);
        System.out.println("输入: nums = [1,2,3]");
        System.out.println("输出: " + result3);
        
        // 测试解法三
        System.out.println("\n=== 解法三测试 ===");
        List<List<Integer>> result4 = solution.subsets3(nums1);
        System.out.println("输入: nums = [1,2,3]");
        System.out.println("输出: " + result4);
    }
}

File: class054_DynamicProgramming_SubsequenceAndKnapsackProblems/LeetCode78_Subsets.py
"""
LeetCode 78. 子集
给你一个整数数组 nums ，数组中的元素互不相同。返回该数组所有可能的子集（幂集）。
解集不能包含重复的子集。你可以按任意顺序返回解集。
测试链接：https://leetcode.cn/problems/subsets/

算法详解：
使用多种方法生成数组的所有子集，包括位掩码法、回溯法和迭代法。

时间复杂度：O(n * 2^n)，其中n是数组长度
空间复杂度：O(n * 2^n) 用于存储所有子集

工程化考量：
1. 异常处理：检查输入参数有效性
2. 边界处理：空数组的情况
3. 性能优化：使用生成器避免内存爆炸
4. 代码质量：清晰的变量命名和类型注解

Python特性：
1. 列表推导式简化代码
2. 生成器表达式支持惰性求值
3. 内置函数提高开发效率
4. 动态类型使得代码简洁
"""

from typing import List, Generator
import time
import itertools

class SubsetsSolution:
    """
    子集问题解决方案类
    提供多种算法实现和测试功能
    """
    
    @staticmethod
    def subsets_bitmask(nums: List[int]) -> List[List[int]]:
        """
        位掩码法生成所有子集
        时间复杂度：O(n * 2^n)
        空间复杂度：O(n * 2^n)
        
        Args:
            nums: 输入数组，元素互不相同
            
        Returns:
            List[List[int]]: 所有子集的列表
            
        Raises:
            TypeError: 输入不是列表类型
        """
        if not isinstance(nums, list):
            raise TypeError("输入必须是列表类型")
            
        n = len(nums)
        result = []
        
        # 特殊情况：空数组
        if n == 0:
            return [[]]
            
        # 总子集数：2^n
        total = 1 << n
        
        for mask in range(total):
            subset = []
            for i in range(n):
                if mask & (1 << i):
                    subset.append(nums[i])
            result.append(subset)
            
        return result
    
    @staticmethod
    def subsets_backtrack(nums: List[int]) -> List[List[int]]:
        """
        回溯法生成所有子集
        时间复杂度：O(n * 2^n)
        空间复杂度：O(n) 递归栈空间
        """
        if not isinstance(nums, list):
            raise TypeError("输入必须是列表类型")
            
        result = []
        
        def backtrack(index: int, path: List[int]) -> None:
            # 基本情况：处理完所有元素
            if index == len(nums):
                result.append(path.copy())
                return
                
            # 不包含当前元素
            backtrack(index + 1, path)
            
            # 包含当前元素
            path.append(nums[index])
            backtrack(index + 1, path)
            path.pop()  # 回溯
            
        backtrack(0, [])
        return result
    
    @staticmethod
    def subsets_iterative(nums: List[int]) -> List[List[int]]:
        """
        迭代法生成所有子集
        时间复杂度：O(n * 2^n)
        空间复杂度：O(n * 2^n)
        """
        if not isinstance(nums, list):
            raise TypeError("输入必须是列表类型")
            
        result = [[]]  # 初始包含空集
        
        for num in nums:
            # 为每个已有子集创建新子集并添加当前元素
            new_subsets = []
            for subset in result:
                new_subset = subset + [num]
                new_subsets.append(new_subset)
            result.extend(new_subsets)
            
        return result
    
    @staticmethod
    def subsets_optimized(nums: List[int]) -> List[List[int]]:
        """
        优化的回溯法（避免不必要的拷贝）
        时间复杂度：O(n * 2^n)
        空间复杂度：O(n) 递归栈空间
        """
        if not isinstance(nums, list):
            raise TypeError("输入必须是列表类型")
            
        result = []
        
        def dfs(start: int, path: List[int]) -> None:
            # 每次递归都将当前路径加入结果
            result.append(path.copy())
            
            for i in range(start, len(nums)):
                path.append(nums[i])
                dfs(i + 1, path)
                path.pop()  # 回溯
                
        dfs(0, [])
        return result
    
    @staticmethod
    def subsets_builtin(nums: List[int]) -> List[List[int]]:
        """
        使用Python内置函数生成子集
        时间复杂度：O(n * 2^n)
        空间复杂度：O(n * 2^n)
        
        注意：这种方法简洁但可能不够直观
        """
        if not isinstance(nums, list):
            raise TypeError("输入必须是列表类型")
            
        result = []
        n = len(nums)
        
        # 使用内置组合函数生成所有子集
        for k in range(n + 1):
            for subset in itertools.combinations(nums, k):
                result.append(list(subset))
                
        return result
    
    @staticmethod
    def subsets_generator(nums: List[int]) -> Generator[List[int], None, None]:
        """
        生成器版本，支持惰性求值
        适用于大规模数据，避免内存爆炸
        
        Yields:
            List[int]: 一个个子集
        """
        if not isinstance(nums, list):
            raise TypeError("输入必须是列表类型")
            
        n = len(nums)
        
        for mask in range(1 << n):
            subset = []
            for i in range(n):
                if mask & (1 << i):
                    subset.append(nums[i])
            yield subset
    
    @staticmethod
    def run_tests() -> None:
        """
        运行单元测试，验证算法的正确性
        """
        print("=== LeetCode 78 子集问题测试 ===\n")
        
        test_cases = [
            # (描述, 输入数组)
            ("空数组", []),
            ("单元素", [1]),
            ("双元素", [1, 2]),
            ("三元素", [1, 2, 3]),
            ("四元素", [1, 2, 3, 4]),
        ]
        
        methods = [
            ("位掩码法", SubsetsSolution.subsets_bitmask),
            ("回溯法", SubsetsSolution.subsets_backtrack),
            ("迭代法", SubsetsSolution.subsets_iterative),
            ("优化回溯", SubsetsSolution.subsets_optimized),
            ("内置函数", SubsetsSolution.subsets_builtin),
        ]
        
        all_passed = True
        
        for description, nums in test_cases:
            print(f"{description}:")
            print(f"  输入数组: {nums}")
            
            # 使用第一种方法作为基准
            try:
                baseline = SubsetsSolution.subsets_bitmask(nums)
                expected_size = 1 << len(nums) if nums else 1
                
                print(f"  预期子集数: {expected_size}")
                print(f"  实际子集数: {len(baseline)}")
                
                case_passed = True
                results = [baseline]
                
                for method_name, method in methods[1:]:  # 跳过基准方法
                    try:
                        result = method(nums)
                        results.append(result)
                        
                        # 检查子集数量
                        size_ok = len(result) == len(baseline)
                        # 检查内容（转换为集合比较）
                        content_ok = (set(tuple(sorted(sub)) for sub in result) == 
                                    set(tuple(sorted(sub)) for sub in baseline))
                        
                        status = "✓" if size_ok and content_ok else "✗"
                        print(f"  {method_name}: {len(result)}个子集 {status}")
                        
                        if not (size_ok and content_ok):
                            case_passed = False
                            all_passed = False
                            
                    except Exception as e:
                        print(f"  {method_name}: 错误 - {e}")
                        case_passed = False
                        all_passed = False
                
                if case_passed:
                    # 打印前几个子集作为示例
                    if len(baseline) <= 16:
                        print(f"  所有子集: {baseline}")
                    else:
                        print(f"  前4个子集: {baseline[:4]} ... 等{len(baseline)}个子集")
                    print("  测试通过 ✓")
                else:
                    print("  测试失败 ✗")
                    
            except Exception as e:
                print(f"  基准方法错误: {e}")
                all_passed = False
                
            print()
            
        if all_passed:
            print("所有测试通过！ ✓")
        else:
            print("部分测试失败！ ✗")
            
        print()
    
    @staticmethod
    def performance_test() -> None:
        """
        性能测试，测试算法在大规模数据下的表现
        """
        print("=== 性能测试 ===")
        
        # 生成测试数据：中等规模数组
        n = 15  # 2^15 = 32768个子集
        nums = list(range(1, n + 1))
        
        print(f"测试数据规模: {n}个元素")
        print(f"预期子集数量: {1 << n}")
        print()
        
        methods = [
            ("位掩码法", SubsetsSolution.subsets_bitmask),
            ("优化回溯", SubsetsSolution.subsets_optimized),
            ("内置函数", SubsetsSolution.subsets_builtin),
        ]
        
        results = {}
        
        for method_name, method in methods:
            start_time = time.time()
            result = method(nums)
            end_time = time.time()
            duration = (end_time - start_time) * 1000  # 转换为毫秒
            
            results[method_name] = len(result)
            
            print(f"{method_name}:")
            print(f"  子集数量: {len(result)}")
            print(f"  耗时: {duration:.2f} 毫秒")
            print()
        
        # 测试生成器版本（避免内存爆炸）
        print("生成器版本测试（避免内存爆炸）:")
        start_time = time.time()
        count = 0
        for subset in SubsetsSolution.subsets_generator(nums):
            count += 1
            # 这里可以处理每个子集，但为了性能测试只计数
        end_time = time.time()
        gen_duration = (end_time - start_time) * 1000
        
        print(f"  子集数量: {count}")
        print(f"  耗时: {gen_duration:.2f} 毫秒")
        print("  注意：生成器版本可以处理更大规模的数据")
        print()
        
        # 验证结果一致性
        if len(set(results.values())) == 1 and count == results["位掩码法"]:
            print("结果一致性验证: 通过 ✓")
        else:
            print("结果一致性验证: 失败 ✗")


def main():
    """
    主函数，运行测试和性能测试
    """
    try:
        # 运行单元测试
        SubsetsSolution.run_tests()
        
        # 运行性能测试
        SubsetsSolution.performance_test()
        
        print("所有测试完成！")
        
    except Exception as e:
        print(f"测试过程中发生错误: {e}")
        return 1
        
    return 0


if __name__ == "__main__":
    exit(main())


"""
复杂度分析详细计算：

位掩码法：
- 时间：外层循环2^n次，内层循环n次 → O(n * 2^n)
- 空间：需要存储所有子集 → O(n * 2^n)

回溯法：
- 时间：生成2^n个子集 → O(n * 2^n)
- 空间：递归深度n → O(n)

迭代法：
- 时间：外层循环n次，内层循环2^i次 → O(n * 2^n)
- 空间：需要存储所有子集 → O(n * 2^n)

优化回溯法：
- 时间：O(n * 2^n)
- 空间：递归深度n → O(n)

内置函数法：
- 时间：O(n * 2^n)，但实际性能可能因实现而异
- 空间：O(n * 2^n)

生成器版本：
- 时间：O(n * 2^n)
- 空间：O(n) 每次只生成一个子集

Python特性说明：
1. 列表推导式使得代码非常简洁
2. itertools模块提供了强大的组合功能
3. 生成器支持惰性求值，适合处理大规模数据
4. 动态类型使得代码灵活但需要更多测试

调试技巧：
1. 使用小规模数据验证算法正确性
2. 打印中间结果观察生成过程
3. 使用断言验证关键假设
4. 使用memory_profiler分析内存使用

工程化建议：
1. 对于小规模数据使用任意方法
2. 对于中等规模数据使用优化回溯法
3. 对于大规模数据使用生成器版本
4. 添加类型注解提高代码可读性
5. 编写单元测试覆盖各种边界情况
"""